File: AoC2019/Day11/Day11.cs
using Range = Common.Range;

namespace AoC2019.Day11;

public static partial class Day11 {
    public static void Solve(IEnumerable<string> input) {
        var program = input.Single().ToLongArray();
        Run(program, 0).Count.Part1();

        var painted = Run(program, 1);

        var xRange = Range.FromStartAndEndInclusive(painted.Min(x => x.Key.X), painted.Max(x => x.Key.X));
        var yRange = Range.FromStartAndEndInclusive(painted.Min(x => x.Key.Y), painted.Max(x => x.Key.Y));

        yRange.Select(y =>
                xRange.Select(x => new V(x, y))
                    .Select(v => painted.GetValueOrDefault(v) == 1 ? '█' : ' ')
                    .StringJoin(string.Empty))
            .Prepend(string.Empty)
            .StringJoin("\n")
            .Dump("Part2: ");
    }

    private static Dictionary<V, int> Run(long[] program, int startColor) {
        var current = V.Zero;
        var dir = V.Up;
        var visited = new Dictionary<V, int> { [current] = startColor };
        var computer = new IntCodeComputer(program);
        while (ReadNext(out var instruction)) {
            visited[current] = instruction.Color;
            dir = dir.Rotate(instruction.Turn == 0 ? -90 : 90);
            current += dir;
        }

        return visited;

        bool ReadNext(out (int Color, int Turn) instruction) {
            instruction = (-1, -1);
            computer.AddInput(visited.GetValueOrDefault(current));
            if (!computer.RunToNextOutput()) return false;
            instruction = ((int)computer.Output, (int)computer.GetNextOutput());
            return true;
        }
    }
}

File: src/golang/aoc2017/day11/day11.go
package main

import (
	_ "embed"
	"strings"
)

type V struct {
	col, row int
}

var directions = map[string]V{
	"n":  {0, -2},
	"ne": {1, -1},
	"se": {1, 1},
	"s":  {0, 2},
	"sw": {-1, 1},
	"nw": {-1, -1},
}

//go:embed input.txt
var input string

func main() {
	part1, part2 := solve(input)
	println("part1 =", part1)
	println("part2 =", part2)
}

func solve(s string) (int, int) {
	zero := V{0, 0}
	current := V{0, 0}
	maxDistance := 0

	for _, dir := range strings.Split(s, ",") {
		current.row += directions[dir].row
		current.col += directions[dir].col
		maxDistance = max(maxDistance, distance(zero, current))
	}

	return distance(zero, current), maxDistance
}

func distance(a, b V) int {
	var dcol = abs(a.col - b.col)
	var drow = abs(a.row - b.row)
	return dcol + max(0, (drow-dcol)/2)
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
